torch/csrc/jit/import.cpp

```diff
@@ -366,22 +344,151 @@ ModuleDecoder::ModuleDecoder(
   }
 }
 
+at::ScalarType tensorProtoTypeToATenType(caffe2::TensorProto_DataType data_type) {
+  switch(data_type) {
+    // NB: handle BOOL
+    case caffe2::TensorProto_DataType_UINT8:
+      return at::kByte;
+    case caffe2::TensorProto_DataType_INT8:
+      return at::kChar;
+    case caffe2::TensorProto_DataType_INT16:
+      return at::kShort;
+    case caffe2::TensorProto_DataType_INT32:
+      return at::kInt;
+    case caffe2::TensorProto_DataType_INT64:
+      return at::kLong;
+    case caffe2::TensorProto_DataType_FLOAT16:
+      return at::kHalf;
+    case caffe2::TensorProto_DataType_FLOAT:
+      return at::kFloat;
+    case caffe2::TensorProto_DataType_DOUBLE:
+      return at::kDouble;
+    default:
+      AT_ERROR("Unsupported TensorProto data type");
+  }
+}
+
+class ScriptModuleDeserializer final {
+ public:
+  ScriptModuleDeserializer(const std::string& filename) :
+    ifs_(filename, std::ifstream::in), reader_(&ifs_) {
+      // TODO appropriate support for mmap, right now still use stream reader
+      //std::cout << "filename: " << filename << std::endl;
+  }
+  ScriptModuleDeserializer(std::istream* is) : ifs_(), reader_(is) {}
+  void deserialize(ModuleLookup module_lookup) {
+    torch::ModelDef model_def;
+    at::DataPtr data_ptr;
+    size_t data_size;
+    std::tie(data_ptr, data_size) = reader_.getLastRecord();
+    AT_ASSERTM(model_def.ParseFromArray(data_ptr.get(), data_size),
+        "parse metadata (i.e., ModelDef) failed.");;
+    //std::cout << "loaded model_def: " << model_def.DebugString() << std::endl;
+    moduleLookup_ = module_lookup;
+    const auto& module_def = model_def.main_module();
+    collectParamsInfo(module_def, module_def.name());
+    std::shared_ptr<script::Module> module = moduleLookup_(moduleStack_);
```

Comment:
add a TODO that it can be simplified after better C++/Python interop lands and the list of submodules would be the same